mlem/contrib/kubernetes/base.py

```diff
@@ -0,0 +1,255 @@
+)
+from .build import build_k8s_docker
+from .context import K8sYamlBuildArgs, K8sYamlGenerator
+from .utils import create_k8s_resources, namespace_deleted, pod_is_running
+
+POD_STATE_MAPPING = {
+    "Pending": DeployStatus.STARTING,
+    "Running": DeployStatus.RUNNING,
+    "Succeeded": DeployStatus.STOPPED,
+    "Failed": DeployStatus.CRASHED,
+    "Unknown": DeployStatus.UNKNOWN,
+}
+
+
+class K8sDeploymentState(DeployState):
+    """DeployState implementation for Kubernetes deployments"""
+
+    type: ClassVar = "kubernetes"
+
+    image: Optional[DockerImage] = None
+    """Docker Image being used for Deployment"""
+    deployment_name: Optional[str] = None
+    """Name of Deployment"""
+
+
+class K8sDeployment(MlemDeployment, K8sYamlBuildArgs):
+    """MlemDeployment implementation for Kubernetes deployments"""
+
+    type: ClassVar = "kubernetes"
+
+    server: Optional[Server] = None
+    """type of Server to use, with options such as FastAPI, RabbitMQ etc."""
+    state_type: ClassVar = K8sDeploymentState
+    """type of state for Kubernetes deployments"""
+    registry: Optional[DockerRegistry] = DockerRegistry()
+    """docker registry"""
+    daemon: Optional[DockerDaemon] = DockerDaemon(host="")
+    """docker daemon"""
+    kube_config_file_path: Optional[str] = None
+    """path for kube config file of the cluster"""
+
+    def load_kube_config(self):
+        config.load_kube_config(
+            config_file=self.kube_config_file_path
+            or os.getenv("KUBECONFIG", default="~/.kube/config")
+        )
+
+    def find_index(self, nodes_list, node_name):
+        for i, each_node in enumerate(nodes_list):
+            if each_node.metadata.name == node_name:
+                return i
+        return -1
+
+    def _get_client(self, state: K8sDeploymentState) -> Client:
+        host, port = None, None
+        self.load_kube_config()
+        service = client.CoreV1Api().list_namespaced_service(self.namespace)
+        try:
+            if self.service_type == "NodePort":
+                port = service.items[0].spec.ports[0].node_port
+                node_name = (
+                    client.CoreV1Api()
+                    .list_namespaced_pod(self.namespace)
+                    .items[0]
+                    .spec.node_name
+                )
+                node_list = client.CoreV1Api().list_node().items
+                node_index = self.find_index(node_list, node_name)
+                address_dict = node_list[node_index].status.addresses
+                for each_address in address_dict:
+                    if each_address.type == "ExternalDNS":
+                        host = each_address.address
+            elif self.service_type == "LoadBalancer":
+                port = service.items[0].spec.ports[0].port
+                ingress = service.items[0].status.load_balancer.ingress[0]
+                host = ingress.hostname or ingress.ip
+            else:
+                raise ValueError(
```

Comment:
this will raise if it's clusterip
create an mlem abc with abstract method get_host_and_port

Reply:
done the mlem abc thing, thinking of raising an error to handle ClusterIP case

Reply:
raised an error, done